### PR TITLE
Refactor in Block producer

### PR DESCRIPTION
**Pull request by anne-laure-s**
This PR updates old commits :

- Split Block_producer.run into smaller functions (splitted in this PR) : https://github.com/MinaProtocol/mina/commit/69561b929eef6ee8ba47ed6814ab534b47e3be1e
- Refactor block production control flow : https://github.com/MinaProtocol/mina/commit/93f645732cb3b54c9af8a672f969ca94badbb1eb
- Remove Interruptible usage from block production : https://github.com/MinaProtocol/mina/commit/8ae33e5bd13ae4a01445c39f69ecb7a643612eee
- Debugging block creation schedule : https://github.com/MinaProtocol/mina/commit/2cd1275c04d385ffb3ff4fd7d51d0a43daad6074
- Refactor genesis block creation : https://github.com/MinaProtocol/mina/commit/8b0d8bbd69a38455238cc40df9d69b9c92acf7a1

### CODE REVIEW

**Comment by anne-laure-s** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-nightly-me

File: src/lib/block_producer/block_producer.ml [hunk 1]
```diff
@@ -365,9 +365,8 @@ let handle_block_production_errors ~logger ~rejected_blocks_logger
         ; ("prev_state", Protocol_state.value_to_yojson previous_protocol_state)
         ; ("prev_state_proof", Proof.to_yojson previous_protocol_state_proof)
         ; ("next_state", Protocol_state.value_to_yojson protocol_state)
-          (* Commented out because for large blocks it's an oversized log *)
-          (* ; ( "internal_transition"
-             , Internal_transition.to_yojson internal_transition ) *)
+        ; ( "internal_transition"
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml:368]: This change was probably made by accident

**Review by georgeee — CHANGES_REQUESTED** [review]

**Comment by georgeee** [comment]
Suggest splitting this PR: commits before _[Refactor block production control flow](https://github.com/MinaProtocol/mina/pull/16181/commits/cd20ac108c4daf471a74e169cc6606ea997e266b)_ are extremely trivial movements of code and could be merged as such (we don't even need to test these).

Everything further actually changes control flow, so it's probably better to consider them separately.

File: src/lib/block_producer/block_producer.ml [hunk 2]
```diff
@@ -1235,19 +1153,16 @@ let run ~context:(module Context : CONTEXT) ~vrf_evaluator ~prover ~verifier
...
         | None ->
             log_bootstrap_mode ~logger () ;
-            don't_wait_for
-              (let%map () =
-                 Broadcast_pipe.Reader.iter_until frontier_reader
-                   ~f:(Fn.compose Deferred.return Option.is_some)
-               in
-               check_next_block_timing slot i () )
+            let%map () =
+              Broadcast_pipe.Reader.iter_until frontier_reader
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml:1096]: Add comment: iterates until there is some frontier

File: src/lib/block_producer/block_producer.ml [hunk 3]
```diff
@@ -1295,43 +1210,52 @@ let run ~context:(module Context : CONTEXT) ~vrf_evaluator ~prover ~verifier
...
+                  Ivar.fill ivar () ) ;
+              let intr_ivar = Ivar.create () in
+              let this_step = ref (Some intr_ivar) in
+              let produce_intr =
+                let%map.Interruptible x = produce intr_ivar data in
+                this_step := None ;
+                x
+              in
+              let%map _ = Interruptible.force produce_intr in
+              (* TODO consider uncommenting below or removing interruptible usage completely *)
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml]: This comment is inaccurate (and suggested code has no effect)

Forcing `produce_intr` makes sure that the next iteration won't start before the block is produced. Hence interruptible will not ever be actually interrupted.

**Comment by georgeee** [comment]
Commit _Debugging block creation schedule_ has to be removed from the PR

File: src/lib/block_producer/block_producer.ml [hunk 4]
```diff
@@ -656,441 +547,559 @@ let validate_genesis_protocol_state_block ~genesis_state_hash (b, v) =
...
+            [%log' debug rejected_blocks_logger] ~metadata msg ;
+            [%log fatal] ~metadata msg ;
+            return ()
+      in
+      let%bind res = emit_breadcrumb () in
+      let span = Block_time.diff (Block_time.now time_controller) start in
+      handle_block_production_errors ~logger ~rejected_blocks_logger
+        ~time_taken:span ~previous_protocol_state ~protocol_state res
+
+let iteration ~schedule_next_check ~dispatch_now ~schedule_dispatch
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml:905]: Rename:

* `~schedule_next_check` to `~schedule_next_vrf_check`
* `~dispatch_now` to `~produce_block_now`
* `~next_check_now` to `~next_vrf_check_now`
* `~schedule_dispatch` to `~schedule_block_production`

File: src/lib/block_producer/block_producer.ml [hunk 5]
```diff
@@ -656,441 +547,559 @@ let validate_genesis_protocol_state_block ~genesis_state_hash (b, v) =
...
+    ~next_check_now ~context:(module Context : CONTEXT) ~vrf_evaluator
+    ~time_controller ~coinbase_receiver ~set_next_producer_timing ~frontier
+    ~vrf_evaluation_state ~epoch_data_for_vrf ~ledger_snapshot i slot =
+  O1trace.thread
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml:909]: Remove this `O1trace.thread`, it was introduced for debugging

File: src/lib/block_producer/block_producer.ml [hunk 6]
```diff
@@ -1241,6 +1225,9 @@ let run ~context:(module Context : CONTEXT) ~vrf_evaluator ~prover ~verifier
 let run_precomputed ~context:(module Context : CONTEXT) ~verifier ~trust_system
     ~time_controller ~frontier_reader ~transition_writer ~precomputed_blocks =
   let open Context in
+  let log_bootstrap_mode () =
```
Comment (georgeee) [inline src/lib/block_producer/block_producer.ml]: why it was re-introduced locally?

**Comment by georgeee** [comment]
Last few commits are better be split to standalone PRs.

Point of commit _[Refactor block production control flow](https://github.com/MinaProtocol/mina/pull/16181/commits/cd20ac108c4daf471a74e169cc6606ea997e266b)_ is that this is only a rewrite of previous code: singleton scheduler is only used to schedule one routine to happen strictly after previous scheduled routine was completed, and `Singleton_supervisor`'s parts were accurately integrated into `run`'s code directly.

Commit _[Remove Interruptible usage from block production](https://github.com/MinaProtocol/mina/pull/16181/commits/036e0f035a8efab1d17c147f98b509ee166ee897)_ removes the use of interruptible which becomes obviously redundant (interruptible will never get to be interrupted) after previous commit.

Finally, _[Refactor genesis block creation](https://github.com/MinaProtocol/mina/pull/16181/commits/208bb9b34cd47496c526e24d382054eb461b65af)_ is the only change that modifies behavior (not just refactors the code): instead of trying to defer creation of genesis block to the latest possible point, it's always being executed at the beginning of `run()`.

File: src/lib/block_producer/block_producer.ml (hunk 3 above)
Comment (anne-laure-s) [reply src/lib/block_producer/block_producer.ml]: Is there anything you want to do about this in this MR ?

File: src/lib/block_producer/block_producer.ml (hunk 1 above)
Comment (anne-laure-s) [reply src/lib/block_producer/block_producer.ml]: I reverted this change

File: src/lib/block_producer/block_producer.ml (hunk 6 above)
Comment (anne-laure-s) [reply src/lib/block_producer/block_producer.ml]: Nice catch, it was a mistake, I removed it

**Comment by anne-laure-s** [comment]
I splitted this PR in 4 :
- https://github.com/MinaProtocol/mina/pull/16222 for the first commits
- https://github.com/MinaProtocol/mina/pull/16223 for refactor block production control flow
- https://github.com/MinaProtocol/mina/pull/16224 for remove Interruptible usage from block production
- https://github.com/MinaProtocol/mina/pull/16225 for refactor genesis block creation

File: src/lib/block_producer/block_producer.ml (hunk 3 above)
Comment (georgeee) [reply src/lib/block_producer/block_producer.ml]: Probably not. Maybe it;'s worth copying as a code comment. Here I mostly explained what happens to myself 🗡️ 

CC @anne-laure-s

Comment (georgeee) [reply src/lib/block_producer/block_producer.ml]: Ah, this TODO is removed in PR chain, so no need to port it to code

**Pull request closed without merging**